staging/kos/README.md

```diff
@@ -188,6 +196,119 @@ remote NetworkAttachment node. The flows installed in that bridge will
 make it possible to deliver the packet to the NetworkAttachment's interface.
 
 
+## The Subnets Validator
+
+This is a singleton that attempts to solve the problem of enforcing the
+two following invariants:
+
+  1. There are no subnets with the same VNI and overlapping CIDR blocks.
+  2. There are no subnets with the same VNI and different Kubernetes API
+     namespaces.
+
+The first is intrinsic in the nature of virtual networks and subnets,
+while the second is a consequence of the design choice of confining VXLAN
+virtual networks to a single Kubernetes API namespace.
+
+The general problem here is enforcement of multi-objects invariants. For
+such a simple SDN as `kos`, we could have ignored it, leaving it to the
+user or to a higher-level UI. Since we believe this is a relevant problem
+in distributed systems and we're building `kos` to assess Kubernetes API
+machinery suitability for building distributed systems other than
+Kubernetes itself, we chose to tackle it.
+
+A request to create a subnet *X* with VNI *V* which violates any of the
+invariants should fail. The only way to get this behavior is having the
+extension API server do the validation, because it is the only component
+to see *X* **before** it has already been created. But this alone is not
+reliable. When *X* is being validated, all the other existing subnets with
+VNI *V* must be retrieved, and each one of them must be compared to *X*
+to check if an invariant is violated. But different API objects are not
+necessarily validated sequentially by the extension API server. If a
+subnet *Y* in *conflict* with *X* (i.e, *X* and *Y* mutual existences
+lead to a violation of one invariant) is validated in parallel with *X*,
+they will both be created. This case must be handled, but it is a rare
+occurrence, so we chose to keep the validation by the API server
+nevertheless.
+
+To solve the problem also in the cases where two conflicting subnets
+are created, we replaced the hard-to-enforce constraint that two subnets
+violating the invariants cannot exist with the easier-to-enforce constraint
+that if two subnets violating the invariants exist, subnet consumers use
+only one and ignore the other. To achieve this, each subnet has a `bool`
```

Comment:
s/only one and ignore the other/at most one/

Reply:
Fixed in 75fc302.